src/Compilers/Core/Portable/PEWriter/MetadataHeapsBuilder.cs

```diff
@@ -269,6 +269,14 @@ public int GetUserStringToken(string str)
                 Debug.Assert(!_streamsAreComplete);
 
                 index = _userStringWriter.Position + _userStringHeapStartOffset;
+
+                // User strings are referenced by metadata tokens (8 bits of which are used for the token type) leaving only 24 bits for the offset. 
+                if ((index & 0xFF000000) != 0)
```

Comment:
Perhaps I haven't traced far enough up the possible call stacks. It looks like right now neither MetadataWriter, nor MetadataHeapsBuilder throw any exceptions besides cancellation and unreachable/invalid-value. That is why reporting an diagnostics is my preference.